python-lib/text_extraction/__init__.py

```diff
@@ -118,6 +118,7 @@ def extract_markdown_chunks(markdown, filename):
     for line in lines:
         stripped_line = line.rstrip()
         if not stripped_line.startswith(("    ", "\t")):
+            # Do not remove tab or 4-spaces as they signal code blocks
```

Comment:
Another comment for code blocks: we should in fact have a similar logic than the ` ``` ` case because to start a code block with a tab or 4-spaces, the preceding line should be empty (except if it is the first line of the document).

So for instance:

```md
Some text hello hello
    I think this is code but in fact this is not

    # Actually this is code
````

Reply:
So in fact we should not delete empty lines that come before a code block delimited by tab/4-spaces

Reply:
Actually this makes me think that if we remove the stripping logic altogether, we might end up with much easier code, wdyt ?

Reply:
I don't think there is a problem, we don't really care about knowing if it's a code block or not, we just want to know if a line is a header or not, andI think lines that start with a tab/4-spaces can never be headers (even if the preceding line is not empty).
In commit [c2f65d6](https://github.com/dataiku/dss-plugin-tesseract-ocr/pull/73/commits/c2f65d6e12a1b1515da958d0563c7ab895194899), I also add empty lines to keep the line breaks inside chunks (before line breaks were removed from chunks).

Reply:
If the only thing we care about is headers, then why do we keep logic around stripping lines ?

Reply:
1, 2 or 3 spaces followed by a # is still a header, so the stripping is useful to find these kind of headers

Reply:
Yep you're right : https://spec.commonmark.org/0.30/#atx-heading

But then I think we should be more explicit by whitelisting accepted headers, and sticking closer to the specs from the link above:
- Starts with 0,1,2, or 3 spaces
- Followed by 1 to 3 `#` in our case
- Followed by at least one space, unless the rest of the line is empty

This means that we should only keep the logic that allows us to determine whether  a line is a header or not (unless we are in a code block wrapped by ` ``` ` or ` ~~~ ` : https://spec.commonmark.org/0.30/#code-fence ), and remove the stripping logic IMO

Reply:
Not sure I understand why we need to change anything, the line `if not stripped_line.startswith(("    ", "\t")):` explicitly says that code blocks and header can start with 0, 1, 2 or 3 spaces (less than 4 spaces), I added a comment to explain that better.
Moreover, using `lstrip` only for lines with less than 4 spaces seems to be quite an elegant solution to check code blocks and headers that can start with 0, 1, 2 or 3 spaces.

In commit [eedac17](https://github.com/dataiku/dss-plugin-tesseract-ocr/pull/73/commits/eedac17d2dd2abb0141d8b2b2add5787659e9c09), I removed the stripping from what is in the output (in line `current_text.append(line)`) and remove the `rstrip`.

For the header that needs to have a space after the #, it was already taken care of (in the condition `len(stripped_line) == header_level or stripped_line[header_level] == " "`).

Reply:
OK fair enough I think we reached an acceptable state now 👍